sbi/neural_nets/embedding_nets/lru.py

```diff
@@ -0,0 +1,464 @@
+        x = torch.stack(x, dim=1)
+
+        # Reverse the temporal oder of the 2nd block, i.e., 2nd direction.
+        if self.bidirectional:
+            x = torch.cat(
+                [
+                    x[:, :, : self.state_dim],
+                    torch.flip(x[:, :, self.state_dim :], dims=[1]),
+                ],
+                dim=-1,
+            )
+
+        # Compute the output (for both directions at the same time).
+        y = (x @ self.C.mT).real + input * self.D
+
+        return y
+
+    def _forward_scan(self, input: Tensor, state: Tensor) -> Tensor:
+        """For details on parallel scan, check discussion in Smith et al (2022).
+
+        Args:
+            input: Sequential data of shape (batch_size, sequence_length,
+                input_dim)
+            state: Initial hidden complex state of the LRU of shape
+                ((batch_size, sequence_length, state_dim),
+                 (batch_size, sequence_length, state_dim)).
+
+        Return:
+            Transformed sequential data of shape (batch_size, sequence_length,
+                input_dim)
+        """
+        # Normalize the input-facing matrix.
+        B_norm = self.B * self.gamma.unsqueeze(dim=-1)
+
+        # Precompute the influence of the inputs one the states.
+        u = input.to(dtype=B_norm.dtype)
+        u_times_B_norm = u @ B_norm.T  # complex-values
+
+        u_times_B_norm[:, 0, : self.state_dim] = u_times_B_norm[
+            :, 0, : self.state_dim
+        ] + (
+            self.lambda_complex[: self.state_dim].view(1, -1)
+            * state[:, : self.state_dim]
+        )
+        lambdas = self.lambda_complex.view(1, 1, -1).expand(u_times_B_norm.shape)
+
+        # Create the elements container, and run the scan.
+        # The result is the complete sequence x.
+        elements = (
+            lambdas[:, :, : self.state_dim],
+            u_times_B_norm[:, :, : self.state_dim],
+        )
+        _, x = associative_scan(
+            binary_operator_diag, elements, dim=1, combine_mode="generic"
+        )
+
+        if self.bidirectional:
+            u_times_B_norm[:, -1, self.state_dim :] = u_times_B_norm[
+                :, -1, self.state_dim :
+            ] + (
+                self.lambda_complex[self.state_dim :].view(1, -1)
+                * state[:, self.state_dim :]
+            )
+            elements = (
+                lambdas[:, :, self.state_dim :],
+                u_times_B_norm[:, :, self.state_dim :],
+            )
+            _, x_bi = associative_scan(
+                binary_operator_diag,
+                elements,
+                dim=1,
+                combine_mode="generic",
+                reverse=True,
+            )
+            x = torch.cat([x, x_bi], dim=-1)
+
+        # Compute the output (for both directions at the same time).
+        y = (x @ self.C.mT).real + input * self.D
+
+        return y
+
+
+def binary_operator_diag(
+    element_i: tuple[Tensor, Tensor],
```

Comment:
Use `Tuple` and add `from typing import Tuple`

Reply:
As you wish